scheduler/reconcile.go

```diff
@@ -32,6 +33,10 @@ const (
 	rescheduleWindowSize = 1 * time.Second
 )
 
+type ReconnectingPicker interface {
+	PickReconnectingAlloc(disconnect *structs.DisconnectStrategy, original *structs.Allocation, replacement *structs.Allocation) *structs.Allocation
+}
```

Comment:
Why do we need this interface? `reconnectingpicker.ReconnectingPicker` seems to be the only implementation.

I could see a case where the drain strategy could be an interface with one implementation per strategy, but `ReconnectingPicker` seems to be handling all of them already.

Reply:
Im aiming at removing the responsibility of the picking from the allocReconciler, that is the main reason of te interface :)

Reply:
Interfaces in Go are a bit strange, and work differently than in other languages, like Java or C#.

Since they're implicitly defined pre-creating them can often lead to a confusing implementation. Specially for people new to the code base, trying to "jump to definition" in an interface is quite frustrating 😬 

Sometimes that's unavoidable because we want to have multiple implementations (like a real thing and a test implementation), or if want to restrict what to expose (like only exposing certain methods of a struct).

But in this case I can't really think of a different implementations of `ReconnectingPicker`. The entire decision is already encapsulated in the `reconnectingpicker.ReconnectingPicker` implementation (the fact they have the same name kind of point to that as well 😅).

So we don't need this interface to remove responsibility from the reconciler. Using the concrete struct already accomplishes that. And having it creates unnecessary indirection that makes the code harder to follow.

If, later on, we do find a need for an interface, the implicit nature of Go's interface can help. If the reconciler ends up calling X(), Y(), and Z(), where each behaves differently under some conditions, we can easily add them to a new interface without much code change.

Reply:
Having that interface in place gives us a define frontier between the alloc reconciler and the reconnecting picker: Imagine we refactor the reconciler and we want to unit test the logic. If we remove the interface, we remove the possibility of having a look at the internals of the reconciliation, we will need to write tests that also take into account the picker logic, no way to isolate components. If we leave the interface where it is, it is the perfect cutting point where we can add a test implementation of the picker just for the purpose  of verifying the behaviour of the reconciliation.

Reply:
It is meant to isolate the logic of the reconciler, not of the picker